servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/HttpMessageDiscardWatchdogServiceTest.java

```diff
@@ -0,0 +1,170 @@
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+import java.util.stream.Stream;
+
+import static io.servicetalk.http.netty.BuilderUtils.newClientBuilder;
+import static io.servicetalk.http.netty.BuilderUtils.newServerBuilder;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+final class HttpMessageDiscardWatchdogServiceTest {
+
+    @RegisterExtension
+    static final ExecutionContextExtension SERVER_CTX =
+            ExecutionContextExtension.cached("server-io", "server-executor")
+                    .setClassLevel(true);
+    @RegisterExtension
+    static final ExecutionContextExtension CLIENT_CTX =
+            ExecutionContextExtension.cached("client-io", "client-executor")
+                    .setClassLevel(true);
+
+    @ParameterizedTest(name = "{displayName} [{index}] transformer={0} expectedSubscriptions={1}")
+    @MethodSource("responseTransformers")
+    void cleansPayloadBodyIfDiscardedInFilter(final ResponseTransformer transformer,
+                                              final int expectedSubscriptions) throws Exception {
+        final CountDownLatch payloadSubscriptionCounter = new CountDownLatch(expectedSubscriptions);
+
+        try (HttpServerContext serverContext = newServerBuilder(SERVER_CTX)
+                .appendServiceFilter(service -> new StreamingHttpServiceFilter(service) {
+                    @Override
+                    public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                                final StreamingHttpRequest request,
+                                                                final StreamingHttpResponseFactory responseFactory) {
+                        return transformer.apply(delegate().handle(ctx, request, responseFactory), responseFactory);
+                    }
+                })
+                .listenStreamingAndAwait((ctx, request, responseFactory) -> Single.fromSupplier(() -> {
+                    final Publisher<Buffer> buffer = Publisher
+                            .from(ctx.executionContext().bufferAllocator().fromUtf8("Hello, World!"))
+                            .beforeOnSubscribe(subscription -> payloadSubscriptionCounter.countDown());
+                    return responseFactory.ok().payloadBody(buffer);
+                }))) {
+
+            try (BlockingHttpClient client = newClientBuilder(serverContext, CLIENT_CTX)
+                    .buildBlocking()) {
+                HttpResponse response = client.request(client.get("/"));
+                assertEquals(0, response.payloadBody().readableBytes());
+            }
+
+            assertTrue(payloadSubscriptionCounter.await(5, TimeUnit.SECONDS));
```

Comment:
just fyi: not necessary to explicitly set timeouts because every test has a default timeout of 10 seconds (30 when running on CI)